### PR TITLE
Fix PAPI counters

### LINKED ISSUE #4123: PAPI performance counters are inaccessible
## Expected Behavior
I have built HPX with support for the PAPI performance counters. The resulting collection of HPX libraries includes `libhpx_papi_counters.so`.

According to the documentation:

> For a full list of available PAPI events and their (short) description
> use the --hpx:list-counters and --papi-event-info=all command line options.

When using these options, I expect to see a list of performance counters including the names of PAPI performance counters.

## Actual Behavior
The latter option is not recognized though:

```
my_exe --hpx:list-counters --papi-event-info=all > out
runtime_support::load_components: command line processing: unrecognised
option '--papi-event-info=all'
```

A list of performance counters is printed to the screen, but does not include the names of PAPI performance counters.

## Steps to Reproduce the Problem
  1. I use CMake's FetchContent module to incorporate HPX in my project and I 
don't perform an install step. My project's targets and HPX's targets 
are put in the same bin and lib build directories. See below for the HPX configuration.
  1. Pick an HPX exe and ask for the performance counters, including the PAPI ones

## Specifications
```
{config}:
  HPX_WITH_AGAS_DUMP_REFCNT_ENTRIES=OFF
  HPX_WITH_APEX=OFF
  HPX_WITH_ATTACH_DEBUGGER_ON_TEST_FAILURE=OFF
  HPX_WITH_AUTOMATIC_SERIALIZATION_REGISTRATION=ON
  HPX_WITH_CXX14_RETURN_TYPE_DEDUCTION=TRUE
  HPX_WITH_DEPRECATION_WARNINGS=ON
  HPX_WITH_GOOGLE_PERFTOOLS=ON
  HPX_WITH_HPXMP=OFF
  HPX_WITH_IO_COUNTERS=ON
  HPX_WITH_IO_POOL=ON
  HPX_WITH_ITTNOTIFY=OFF
  HPX_WITH_LOGGING=ON
  HPX_WITH_MORE_THAN_64_THREADS=ON
  HPX_WITH_NATIVE_TLS=ON
  HPX_WITH_NETWORKING=ON
  HPX_WITH_PAPI=ON
  HPX_WITH_PARCELPORT_ACTION_COUNTERS=ON
  HPX_WITH_PARCELPORT_LIBFABRIC=OFF
  HPX_WITH_PARCELPORT_MPI=ON
  HPX_WITH_PARCELPORT_MPI_MULTITHREADED=ON
  HPX_WITH_PARCELPORT_TCP=ON
  HPX_WITH_PARCELPORT_VERBS=OFF
  HPX_WITH_PARCEL_COALESCING=ON
  HPX_WITH_PARCEL_PROFILING=OFF
  HPX_WITH_SANITIZERS=OFF
  HPX_WITH_SCHEDULER_LOCAL_STORAGE=OFF
  HPX_WITH_SPINLOCK_DEADLOCK_DETECTION=OFF
  HPX_WITH_STACKTRACES=ON
  HPX_WITH_SWAP_CONTEXT_EMULATION=OFF
  HPX_WITH_TESTS_DEBUG_LOG=OFF
  HPX_WITH_THREAD_BACKTRACE_ON_SUSPENSION=OFF
  HPX_WITH_THREAD_CREATION_AND_CLEANUP_RATES=OFF
  HPX_WITH_THREAD_CUMULATIVE_COUNTS=ON
  HPX_WITH_THREAD_DEBUG_INFO=OFF
  HPX_WITH_THREAD_DESCRIPTION_FULL=OFF
  HPX_WITH_THREAD_GUARD_PAGE=ON
  HPX_WITH_THREAD_IDLE_RATES=ON
  HPX_WITH_THREAD_LOCAL_STORAGE=OFF
  HPX_WITH_THREAD_MANAGER_IDLE_BACKOFF=ON
  HPX_WITH_THREAD_QUEUE_WAITTIME=OFF
  HPX_WITH_THREAD_STACK_MMAP=ON
  HPX_WITH_THREAD_STEALING_COUNTS=OFF
  HPX_WITH_THREAD_TARGET_ADDRESS=OFF
  HPX_WITH_TIMER_POOL=ON
  HPX_WITH_TUPLE_RVALUE_SWAP=ON
  HPX_WITH_VALGRIND=OFF
  HPX_WITH_VERIFY_LOCKS=OFF
  HPX_WITH_VERIFY_LOCKS_BACKTRACE=OFF
  HPX_WITH_VERIFY_LOCKS_GLOBALLY=OFF

  HPX_PARCEL_MAX_CONNECTIONS=512
  HPX_PARCEL_MAX_CONNECTIONS_PER_LOCALITY=4
  HPX_AGAS_LOCAL_CACHE_SIZE=4096
  HPX_HAVE_MALLOC=tcmalloc
  HPX_PREFIX (configured)=
  HPX_PREFIX=/quanta1/home/jong0137/development/objects/RelWithDebInfo/lue
{version}: V1.3.0 (AGAS: V3.0), Git: a943fd2c5d
{boost}: V1.65.1
{build-type}: release
{date}: Sep 24 2019 15:35:21
{platform}: linux
{compiler}: GNU C++ version 7.2.0
{stdlib}: GNU libstdc++ version 20170814
```

### LINKED ISSUE #4820: PAPI counters don't work
Unsure if this is a system, PAPI, APEX, or HPX issue, so looking for input from e.g. @khuck and @biddisco.

Currently our PAPI tests fail like this on Piz Daint: https://cdash.cscs.ch/testDetails.php?test=23719608&build=115247. As far as I know nothing has changed in our PAPI counters for a long while (except build-related options). That particular example asks for `PAPI_FP_INS` which `--hpx:papi-event-info=all` does not list, `papi_avail` also lists it as not available, so in that sense the failure is reasonable. However, I've tried with other counters that are listed as available and the same error appears. The exception is thrown by the PAPI performance counters in HPX but it seems like APEX is unhappy with something before that as well.

Has anyone tried the PAPI counters recently? This does smell like it's just a problem with PAPI on Piz Daint.

### PR DESCRIPTION
**Pull request by msimberg**
Fixes #4820. It seems like PAPI was not happy with the changes to thread ids in #4693. From my testing the only thread ids that PAPI accepts are the ones from the `gettid` syscall. This was (and is) Linux only, so I've also added a check that the PAPI counters can only be enabled on Linux (as far as I can tell they're definitely not available on Windows; other unixy platforms may support it but that would need testing by someone on those platforms).

Other changes:
- I've disabled one of the PAPI tests as it relies on the `PAPI_FP_INS` counter which doesn't seem to be available everywhere (I'm hoping the two other tests will be enough to make sure we don't completely break the PAPI counters).
- Ignoring a lock that's held if an error happens during counter construction.
- Use `CMAKE_INSTALL_LIBDIR` for component paths (`lib64` instead of `lib` on some systems). I think this fixes #4123.

### CODE REVIEW

**Comment by msimberg** [comment]
retest

**Comment by msimberg** [comment]
@hkaiser could I ask you to help me out with this: https://github.com/STEllAR-GROUP/hpx/runs/894205983?check_suite_focus=true#step:5:20? I assume I'm generating something bad with this: https://github.com/STEllAR-GROUP/hpx/pull/4845/files#diff-af3b638bc2a3e6c650974192a53c7291R2264-R2269, but I can't tell what it would be, and I can't get MSVC to print something useful for me (`-verbosity:detailed` sure enough produced more output, but it didn't want to print the offending line).

**Comment by hkaiser** [comment]
> @hkaiser could I ask you to help me out with this: https://github.com/STEllAR-GROUP/hpx/runs/894205983?check_suite_focus=true#step:5:20? I assume I'm generating something bad with this: https://github.com/STEllAR-GROUP/hpx/pull/4845/files#diff-af3b638bc2a3e6c650974192a53c7291R2264-R2269, but I can't tell what it would be, and I can't get MSVC to print something useful for me (`-verbosity:detailed` sure enough produced more output, but it didn't want to print the offending line).

The cmake code you linked produces:
```
#if !defined(/bin/hpx")
#define /bin/hpx"
#endif
``` 
which probably not what you wanted... I think cmake is interpreting the `';'` as a list delimiter and gets confused by the whole construct. I have pushed a possible fix to this branch (hope you don't mind).

**Comment by msimberg** [comment]
> > @hkaiser could I ask you to help me out with this: https://github.com/STEllAR-GROUP/hpx/runs/894205983?check_suite_focus=true#step:5:20? I assume I'm generating something bad with this: https://github.com/STEllAR-GROUP/hpx/pull/4845/files#diff-af3b638bc2a3e6c650974192a53c7291R2264-R2269, but I can't tell what it would be, and I can't get MSVC to print something useful for me (`-verbosity:detailed` sure enough produced more output, but it didn't want to print the offending line).
> 
> The cmake code you linked produces:
> 
> ```
> #if !defined(/bin/hpx")
> #define /bin/hpx"
> #endif
> ```
> 
> which is probably not what you wanted... I think cmake is interpreting the `';'` as a list delimiter and gets confused by the whole construct. I have pushed a possible fix to this branch (hope you don't mind).

Don't mind at all, thank you! This is exactly what I was missing.

File: libs/runtime_configuration/src/runtime_configuration.cpp [hunk 1]
```diff
@@ -73,6 +73,21 @@ namespace hpx { namespace parcelset {
 }}    // namespace hpx::parcelset
 
 namespace hpx { namespace util {
+
+    namespace detail {
+
+        // CMake does deal with explicit semicolons well, for this reason, the
```
Comment (hkaiser) [inline libs/runtime_configuration/src/runtime_configuration.cpp:79]: I most likely meant 'does _not_ deal well' *grumble*

Comment (msimberg) [reply libs/runtime_configuration/src/runtime_configuration.cpp:79]: I think you clearly wrote that comment sarcastically ;)

**Pull request merged**